### ISSUE TITLE
Ubuntu 14.04 swift 3.1 problem

### ISSUE THREAD

**Issue by savelii** [OpenKitten/MongoKitten#103]
Hi, I am trying to use MongoKitten on IBM Bluemix and local machine, both Ubuntu 14.04 with two swift 3.1 latest snapshots (swift-3.1-DEVELOPMENT-SNAPSHOT-2017-03-15-a-ubuntu14.04 and swift-3.1-DEVELOPMENT-SNAPSHOT-2017-03-19-a-ubuntu14.04). 

I used little modified branch of BSON and MongoKitten (RegularExpression changed to NSRegularExpression in swift 3.1 thats all changes, nothing special) you can look here 
https://github.com/savelii/BSON
https://github.com/savelii/MongoKitten (forked from https://github.com/lgaches/MongoKitten)

Tried with mLab and localhost same error, falling on  try Database(mongoURL:... line, and doesn't even jump in catch block
```
print("Checking if database is up to date")
let database: Database!

do {
    //database = try Database(mongoURL: "mongodb://***:***@ds***.mlab.com:31320/presentation")
    database = try Database(mongoURL: "mongodb://***:***@localhost:27017/admin?ssl=true")
    if database.server.isConnected {
        print("Connected successfully to server")
    }
} catch {
    print("Cannot connect to MongoDB")
    exit(1)
}
```

error from blue mix:
`APP/0*** Error in 'test-app': corrupted double-linked list: 0x00000000009761e0 ***`

error on my local machine when connect to localhost and sometimes to mLab:
```
saveliy@ubuntu:~/test/TestApp$ ./.build/debug/test-app
Checking if database is up to date
test-app: malloc.c:2372: sysmalloc: Assertion `(old_top == (((mbinptr) (((char *) &((av)->bins[((1) - 1) * 2])) - __builtin_offsetof (struct malloc_chunk, fd)))) && old_size == 0) || ((unsigned long) (old_size) >= (unsigned long)((((__builtin_offsetof (struct malloc_chunk, fd_nextsize))+((2 *(sizeof(size_t))) - 1)) & ~((2 *(sizeof(size_t))) - 1))) && ((old_top)->size & 0x1) && ((unsigned long) old_end & pagemask) == 0)' failed.
Aborted (core dumped)
```
error on my local machine when connect to mLab a bit different each time:
```
saveliy@ubuntu:~/test/TestApp$ ./.build/debug/test-app
Checking if database is up to date
*** Error in `./.build/debug/test-app': munmap_chunk(): invalid pointer: 0x0000000001351db0 ***
Aborted (core dumped)
saveliy@ubuntu:~/test/TestApp$ ./.build/debug/test-app
Checking if database is up to date
*** Error in `./.build/debug/test-app': munmap_chunk(): invalid pointer: 0x0000000002d5eea0 ***
Aborted (core dumped)
saveliy@ubuntu:~/test/TestApp$ ./.build/debug/test-app
Checking if database is up to date
test-app: malloc.c:2372: sysmalloc: Assertion `(old_top == (((mbinptr) (((char *) &((av)->bins[((1) - 1) * 2])) - __builtin_offsetof (struct malloc_chunk, fd)))) && old_size == 0) || ((unsigned long) (old_size) >= (unsigned long)((((__builtin_offsetof (struct malloc_chunk, fd_nextsize))+((2 *(sizeof(size_t))) - 1)) & ~((2 *(sizeof(size_t))) - 1))) && ((old_top)->size & 0x1) && ((unsigned long) old_end & pagemask) == 0)' failed.
Aborted (core dumped)
saveliy@ubuntu:~/test/TestApp$ ./.build/debug/test-app
Checking if database is up to date
*** Error in `./.build/debug/test-app': free(): invalid size: 0x000000000286fdb0 ***
Aborted (core dumped)
saveliy@ubuntu:~/test/TestApp$ ./.build/debug/test-app
Checking if database is up to date
Segmentation fault (core dumped)
```

So I ready to spent some time investigating and searching solution, my question is maybe your already familiar with this issue? 
Is there any branch suppose to work with swift 3.1 snapshots? 
I want to understand from where to start? There is 3.x.x 4.x.x and 5.x.x releases, which of them is most likely close to work with swift 3.1 or maybe you already has tested some branch?

**Comment by savelii** [comment]
Just tested with MongoKitten 3.1.0-beta2 it works fine with Swift 3.0.2, so I just make quick changes (RegularExpression changed to NSRegularExpression) for quick test and again got runtime errors after compile with Swift 3.1

much same errors:
```
saveliy@ubuntu:~/test/TestApp$ ./.build/debug/test-app
Checking if database is up to date
*** Error in `./.build/debug/test-app': munmap_chunk(): invalid pointer: 0x000000000141e1a0 ***
Aborted (core dumped)
saveliy@ubuntu:~/test/TestApp$ ./.build/debug/test-app
Checking if database is up to date
*** Error in `./.build/debug/test-app': free(): invalid size: 0x0000000002d8c1f0 ***
Aborted (core dumped)
saveliy@ubuntu:~/test/TestApp$ ./.build/debug/test-app
Checking if database is up to date
Segmentation fault (core dumped)
saveliy@ubuntu:~/test/TestApp$ ./.build/debug/test-app
Checking if database is up to date
Segmentation fault (core dumped)
saveliy@ubuntu:~/test/TestApp$ ./.build/debug/test-app
Checking if database is up to date
*** Error in `./.build/debug/test-app': free(): invalid size: 0x00000000011fc1f0 ***
Aborted (core dumped)
saveliy@ubuntu:~/test/TestApp$ ./.build/debug/test-app
Checking if database is up to date
*** Error in `./.build/debug/test-app': corrupted double-linked list: 0x0000000001ab2460 ***
Aborted (core dumped)
```

**Comment by Joannis** [comment]
I've tried investigating this but it seems that it's probably an issue with the fork by @lgaches using BlueSSL, using no SSL and MLab works fine using your fork.

If you don't need BlueSSL but vapor's LibreSSL works, you can use the bson5 branch. More information on that is [available here.](http://beta.openkitten.org)

MongoKitten 4 is under development, so locking onto a beta tag is advised. The API will change a little over the next weeks. Most, if not all of your code should still compile and work as expected if you upgrade to MongoKitten 4's final release which we'll release just after 3.1 is released.

**Comment by lgaches** [comment]
MongoKitten 3.x hasn't be tested with Swift 3.1. If you are using Kitura, MongoKitten with Vapor's LibreSSL, It's not stable. 
So my advise is if you are using Kitura use my fork and Swift 3.0.2 it's working fine. When Swift 3.1 is released, I'll support MongoKitten 4 with Kitura stack.

**Comment by savelii** [comment]
> I've tried investigating this but it seems that it's probably an issue with the fork by @lgaches using BlueSSL, using no SSL and MLab works fine using your fork.
Can you describe how you get my fork work with "no SSL and MLab"? You modified my fork by removing some BlueSSL dependencies? or just some special way to init database (another then database = try Database(mongoURL: "mongodb://***:***@ds***.mlab.com:31320/presentation"))? I can't get it worked in any way. Thanks.

**Comment by Joannis** [comment]
I just used the MLab connection string without `?ssl=true`

**Comment by savelii** [comment]
Very strange, it's keep falling for me with swift-3.1-DEVELOPMENT-SNAPSHOT-2017-03-19-a-ubuntu14.04. 

By the way, just for info, noticed strange behaviour with 
`.Package(url: "https://github.com/lgaches/MongoKitten.git", "3.1.0-beta3")`
it downloads not package with exact this tag, it downloads older package ignoring "-beta3" part. Same with "5.0.0-beta1" SPM resolves and fetch "5.0.0-obbut2" for me. Looks like SPM bug.

Ok, anyway will stuck with Swift 3.0.2 for now, it's ok for me. I think this can be closed, thanks.
